File: services/index/internal/chunker/chunker.go
package chunker

import (
	"strings"
)

type Config struct {
	ChunkSize    int
	ChunkOverlap int
	MaxChunks    int
}

type Chunk struct {
	Text     string
	Ordinal  int
	Metadata map[string]string
}

func NewDefaultConfig() Config {
	return Config{ChunkSize: 700, ChunkOverlap: 80, MaxChunks: 0}
}

func SplitText(cfg Config, text string) []Chunk {
	if cfg.ChunkSize <= 0 {
		return nil
	}
	runes := []rune(strings.TrimSpace(text))
	if len(runes) == 0 {
		return nil
	}
	var chunks []Chunk
	step := cfg.ChunkSize - cfg.ChunkOverlap
	if step <= 0 {
		step = cfg.ChunkSize
	}
	for idx, start := 0, 0; start < len(runes); idx, start = idx+1, start+step {
		end := start + cfg.ChunkSize
		if end > len(runes) {
			end = len(runes)
		}
		chunk := strings.TrimSpace(string(runes[start:end]))
		if chunk == "" {
			continue
		}
		chunks = append(chunks, Chunk{Text: chunk, Ordinal: idx})
		if cfg.MaxChunks > 0 && len(chunks) >= cfg.MaxChunks {
			break
		}
		if end == len(runes) {
			break
		}
	}
	return chunks
}

File: services/index-python/pipeline/chunker.py
import logging
from typing import Any, Dict, List

from .config import ChunkConfig
from .types import TextChunk


logger = logging.getLogger(__name__)


class ChunkSplitter:
    def __init__(self, cfg: ChunkConfig):
        self.cfg = cfg

    def split(self, text: str, base_metadata: Dict[str, Any]) -> List[TextChunk]:
        trimmed = text.strip()
        if not trimmed or self.cfg.chunk_size <= 0:
            logger.debug("no text or invalid chunk_size=%s", self.cfg.chunk_size)
            return []

        step = self.cfg.chunk_size - self.cfg.chunk_overlap
        if step <= 0:
            step = self.cfg.chunk_size

        chunks: List[TextChunk] = []
        ordinal = 0
        start = 0
        while start < len(trimmed):
            end = min(len(trimmed), start + self.cfg.chunk_size)
            piece = trimmed[start:end].strip()
            if piece:
                metadata = dict(base_metadata)
                metadata["ordinal"] = ordinal
                chunks.append(TextChunk(text=piece, ordinal=ordinal, metadata=metadata))
                ordinal += 1
                if 0 < self.cfg.max_chunks <= len(chunks):
                    logger.debug("reached max_chunks=%s", self.cfg.max_chunks)
                    break
            if end == len(trimmed):
                break
            start += step
        logger.info("split document into %d chunks", len(chunks))
        return chunks
